README.md
# prtflo-clj-str-b64

## Warning!

This library is not meant to be used in actual projects.  If you're looking for a Base64 encoder/decoder library, you'll need to look elsewhere.

## About

This is a simple portfolio project implementing (what I considered to be) an interesting algorithm.  Thanks for checking it out!

This library can currently encode only ASCII text, and makes no attempt at correctly handling unicode.  

## Installation

There is no provided jar for this project.  In order to run it, it will need to be cloned and worked with via REPL.

1.  Check out the project using whatever Git tool.
2.  `lein repl` within the project's folder
3.  Move into the `prtflo-clj-str-b64.core` namespace
4.  You should now be able to use the `encode` and `decode` functions.

## Usage

### Encoding

Limitations
- Input must be an ASCII-only string

```
(encode "Your input here")
```

### Decoding

Limitations
- Input must be a Base64-encoded string

```
(decode "WW91ciBpbnB1dCBoZXJl")
```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-10-11
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-10-11
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/prtflo-clj-str-b64/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/prtflo-clj-str-b64/compare/0.1.0...0.1.1


test/prtflo_clj_str_b64/core_test.clj
(ns prtflo-clj-str-b64.core-test
  (:require [clojure.test :refer :all]
            [prtflo-clj-str-b64.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/prtflo_clj_str_b64/core.clj
(ns prtflo-clj-str-b64.core
  (:require [prtflo-clj-str-b64.encode :as encode]
            [prtflo-clj-str-b64.decode :as decode]
            [prtflo-clj-str-b64.support :as support]))

;-----------------------------------------------------------------
; Forward declarations
;-----------------------------------------------------------------

(declare perform
         validate-encode-input
         validate-decode-input
         is-string
         is-ascii
         is-base64)

;-----------------------------------------------------------------
; Encoders/decoders
;-----------------------------------------------------------------

(defn encode
  "Base64-encode a value.
  
  The value must be:
  - a string
  - ASCII-only"
  [input]
  (perform validate-encode-input encode/encode input))

(defn decode
  "Base64-decode a value.
  
  The value must be:
  - a string
  - Base64-characters only"
  [input]
  (perform validate-decode-input decode/decode input))

(defn perform
  "Perform a given encoding/decoding
  function after validation.
  
  - Runs validate-fnc against input
  - If validate-fnc returns nil, run 
    action-fnc against input; otherwise,
    throw an exception with validate-fnc's
    return value as the message."
  [validate-fnc action-fnc input]
  (let [e (validate-fnc input)]
    (if (nil? e)
      (action-fnc input)
      (throw (Exception. e)))))

;-----------------------------------------------------------------
; Validators
;-----------------------------------------------------------------

(defn validate-encode-input
  "Determine if input can be
  Base64-encoded."
  [input]
  (cond
    (not (is-string input)) "Input must be a string"
    (not (is-ascii  input)) "Input must be ASCII-only"
    :else nil))

(defn validate-decode-input
  "Determine if input can be
  Base64-decoded."
  [input]
  (cond
    (not (is-string input)) "Input must be a string"
    (not (is-base64 input)) "Input is not base64"
    :else nil))

;-----------------------------------------------------------------
; Predicates
;-----------------------------------------------------------------

(defn is-string
  "Determine if input is a string."
  [input]
  (instance? String input))

(defn is-ascii
  "Determine if input contains
  only ASCII characters."
  [input]
  (let [ints         (map int input)
        out-of-range #(> % 128)
        bads         (filter out-of-range ints)]
    (= 0 (count bads))))

(defn is-base64
  "Determine if input is a
  Base64 value."
  [input]
  (let [b64-chars       (map char support/b64-characters)
        chars           (conj b64-chars \=)
        not-in-charset? #(not (some #{%} chars))
        bads            (filter not-in-charset? input)]
    (= 0 (count bads))))

src/prtflo_clj_str_b64/encode.clj
(ns prtflo-clj-str-b64.encode
  (:require [prtflo-clj-str-b64.support :as support]))

;  ENCODE PROCESS EXAMPLE
;  0.   Start:        "hello"
;  1.   As chrs:      ["h" "e" "l" "l" "o"]
;  2.   As ints:      [104 101 108 108 111]
;  3.   As bits:      ["1101000" "1100101" "1101100" "1101100" "1101111"]
;  4.   As p-bits:    ["01101000" "01100101" "01101100" "01101100" "01101111"]
;  5.   b64-bits-raw: ["011010" "000110" "010101" "101100" "011011" "000110" "1111"]
;  6.   b64-bits:     ["011010" "000110" "010101" "101100" "011011" "000110" "111100="]
;  7.   p-b64-bits:   ["00011010" "00000110" "00010101" "00101100" "00011011" "00000110" "00111100="]
;  8.   As b64 ints:  ["26" "6" "21" "44" "27" "6" "60="]
;  9.   As b64 chrs:  ["a" "G" "V" "s" "b" "G" "8="]
;  10.  End:          "aGVsbG8="

;-----------------------------------------------------------------
; Values
;-----------------------------------------------------------------

(def b64-character-lookup 
  "A map for looking up Base64 characters by
  index, given an integer.
  
  {0 A, 1 B, ...}"
  (zipmap (range 0 64) support/b64-characters))

;-----------------------------------------------------------------
; Forward declarations
;-----------------------------------------------------------------

(declare str-to-char-strs
         char-strs-to-ints
         ints-to-bits
         bits-to-padded-bits
         padded-bits-to-b64-bits
         b64-bits-to-padded-b64-bits
         padded-b64-bits-to-full-b64-bits
         full-b64-bits-to-b64-ints
         b64-ints-to-b64-chars)

;-----------------------------------------------------------------
; Encoder
;-----------------------------------------------------------------

(defn encode
  "Base64 encode the given value.
  
  This value must be an acceptable string,
  but this function will not perform 
  validation."
  [input]
  (->> input
       str-to-char-strs
       char-strs-to-ints
       ints-to-bits
       bits-to-padded-bits
       padded-bits-to-b64-bits
       b64-bits-to-padded-b64-bits
       padded-b64-bits-to-full-b64-bits
       full-b64-bits-to-b64-ints
       b64-ints-to-b64-chars
       support/str-join))

;-----------------------------------------------------------------
; STEP 1 - String to [char-strings]
;-----------------------------------------------------------------

(defn- str-to-char-strs
  "Convert a string to a seq of
  characters-as-strings.
  
  \"abc\" -> [\"a\" \"b\" \"c\"]"
  [input]
  (map str input))

;-----------------------------------------------------------------
; STEP 2 - [char-strings] to [ints]
;-----------------------------------------------------------------

(defn- char-strs-to-ints
  "Convert a seq of characters-as-
  strings to a seq of integers.
  
  [\"a\" \"b\" \"c\"] -> [97 98 99]"
  [input]
  (map (comp int first) input))

;-----------------------------------------------------------------
; STEP 3 - [ints] to [unpadded bit-string]
;-----------------------------------------------------------------

(defn- ints-to-bits
  "Converts a seq of integers to a
  seq of binary-strings.
  
  [97] -> [\"1100001\"]"
  [input]
  (map support/int-to-bin-str input))

;-----------------------------------------------------------------
; STEP 4 - [unpadded bit-string] to [padded bit-string]
;-----------------------------------------------------------------

(defn- bits-to-padded-bits
  "Converts a seq of unpadded binary-strings
  to a seq of 8-width padded binary-strings.
  
  [\"1100001\"] -> [\"01100001\"]"
  [input]
  (map #(support/left-pad 8 %) input))

;-----------------------------------------------------------------
; STEP 5 - [padded bit-string] to [unpadded b64-bit-string]
;-----------------------------------------------------------------

(defn- padded-bits-to-b64-bits
  "Converts a seq of padded binary-strings
  to a seq of unpadded Base64-binary-strings.
  
  [\"01100001\"] -> [\"011000\" \"01\"]"
  [input]
  (support/regroup 6 input))

;-----------------------------------------------------------------
; STEP 6 - [unpadded b64-bit-string] to [padded b64-bit-string]
;-----------------------------------------------------------------

(defn- pad-b64-bits
  "Pads a Base64-binary-string to 8
  characters, adding any equals
  signs.
  
  \"01\" -> \"010000==\""
  [input]
  (let [len     (count input)
        to-add  (- 6 len)
        padding (case to-add
                  0 ""
                  2 "00="
                  4 "0000==")]
    (str input padding)))

(defn- b64-bits-to-padded-b64-bits
  "Pads a seq of Base64-binary-strings to
  6 characters, adding any equals signs.
  
  [\"01\"] -> [\"010000==\"]"
  [input]
  (map pad-b64-bits input))

;-----------------------------------------------------------------
; STEP 7 - [padded b64-bit-string] to [full b64-bit-string]
;-----------------------------------------------------------------

(defn- padded-b64-bits-to-full-b64-bit-seq
  "Pads a padded Base64-binary-string
  to 8 characters, adding any equals
  signs
  
  \"010000==\" -> \"00010000==\""
  [input]
  (let [bits (subs input 0 6)
        eqs  (subs input 6)
        padded (support/left-pad 8 bits)]
    (str padded eqs)))

(defn- padded-b64-bits-to-full-b64-bits
  "Pads a seq of padded Base64-binary-
  strings to 8 characters, adding any
  equals signs.
  
  [\"010000==\"] -> [\"00010000==\"]"
  [input]
  (map padded-b64-bits-to-full-b64-bit-seq input))

;-----------------------------------------------------------------
; STEP 8 - [padded b64-bit-string] to [b64-int]
;-----------------------------------------------------------------

(defn- full-b64-bits-to-b64-int
  "Converts a Base64-binary-string to
  a Base64-integer, adding any equals
  signs.
  
  \"00010000==\" -> \"16==\""
  [input]
  (let [bits        (subs input 0 8)
        eqs         (subs input 8)
        integer     (support/bin-str-to-int bits)]
    (str integer eqs)))

(defn- full-b64-bits-to-b64-ints
  "Converts a seq of Base64-binary-strings
  to Base64-integers, adding any equals signs.
  
  [\"00010000==\"] -> [\"16==\"]"
  [input]
  (map full-b64-bits-to-b64-int input))

;-----------------------------------------------------------------
; STEP 9 - [b64-int] to [b64-char]
;-----------------------------------------------------------------

(defn- b64-int-to-b64-char
  "Converts a Base64-integer to a 
  Base64-charcter, adding any equals
  signs.
  
  \"16==\" -> \"Q==\""
  [input]
  (let [int-str   (support/strip-equals input)
        integer   (read-string int-str)
        equals    (support/strip-num input)
        character (get b64-character-lookup integer)]
    (str character equals)))

(defn- b64-ints-to-b64-chars
  "Converts a seq of Base64-integers
  to Base64-characters, adding any
  equals signs.
  
  [\"16==\"] -> [\"Q==\"]"
  [input]
  (map b64-int-to-b64-char input))

src/prtflo_clj_str_b64/support.clj
(ns prtflo-clj-str-b64.support)

;-----------------------------------------------------------------
; Values
;-----------------------------------------------------------------

(def b64-characters 
  "The acceptable Base64 character set" 
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/")

;-----------------------------------------------------------------
; String helpers
;-----------------------------------------------------------------

(defn left-pad
  "Pad a given string to the width specified by len."
  [len input]
  (let [real-len (count input)
        to-add   (- len real-len)
        padding  (clojure.string/join "" (repeat to-add "0"))]
    (str padding input)))

(defn str-join
  "Join a seq together into a string, using \"\" (empty string) 
  as the separator."
  [items]
  (clojure.string/join "" items))

;-----------------------------------------------------------------
; Partitioning
;-----------------------------------------------------------------

(defn regroup
  "Re-partition a seq of binary numbers (strings) into
  partitions of the given size."
  [size input]
  (let [flattened  (clojure.string/join "" input)
        chr-groups (partition size size [] flattened)
        groups     (map #(clojure.string/join "" %) chr-groups)]
    groups))

;-----------------------------------------------------------------
; String helpers
;-----------------------------------------------------------------

(defn strip-equals
  "Strip out any equals-signs from a string."
  [input]
  (clojure.string/replace input #"=" ""))

(defn strip-alphanum
  "Strip out any alphanumeric characters from
  a string."
  [input]
  (clojure.string/replace input #"\w" ""))

(defn strip-num
  "Strip out any numeric characters from 
  a string."
  [input]
  (clojure.string/replace input #"\d" ""))

;-----------------------------------------------------------------
; Conversion helpers
;-----------------------------------------------------------------

(defn bin-str-to-int
  "Convert a binary number (string) to an integer."
  [input]
  (read-string (str "2r" input)))

(defn int-to-bin-str
  "Convert an integer to a binary number (string)"
  [input]
  (Integer/toString input 2))

src/prtflo_clj_str_b64/decode.clj
(ns prtflo-clj-str-b64.decode
  (:require [prtflo-clj-str-b64.support :as support]))

;  DECODE PROCESS EXAMPLE
;  0.  Start:        "aGVsbG8="
;  1.  As b64 chrs:  ["a" "G" "V" "s" "b" "G" "8="]
;  2.  As b64 ints:  ["26" "6" "21" "44" "27" "6" "60="]
;  3.  raw-b64-bits: ["11010" "110" "10101" "101100" "11011" "110" "111100="]
;  4.  pad-b64-bits: ["011010" "000110" "010101" "101100" "011011" "000110" "111100="]
;  5.  fin-b64-bits: ["011010" "000110" "010101" "101100" "011011" "000110" "1111"]
;  6.  bits:         ["01101000" "01100101" "01101100" "01101100" "01101111"]
;  7.  ints:         [104 101 108 108 111]
;  8.  chars:        ["h" "e" "l" "l" "o"]
;  9.  End:          "hello"

;-----------------------------------------------------------------
; Values
;-----------------------------------------------------------------

(def b64-integer-lookup 
  "A map for looking up integers by
  Base64-characters.
  
  {A 0, B 1, ...}"
  (zipmap support/b64-characters (range 0 64)))

;-----------------------------------------------------------------
; Forward declarations
;-----------------------------------------------------------------

(declare str-to-b64-chars
         b64-chars-to-b64-ints
         b64-ints-to-raw-b64-bits
         raw-b64-bits-to-padded-b64-bits
         padded-b64-bits-to-final-b64-bits
         final-b64-bits-to-bits
         bits-to-ints
         ints-to-char-strs)

;-----------------------------------------------------------------
; Decoder
;-----------------------------------------------------------------

(defn decode
  "Base64 decode the given value.
  
  This value must be an acceptable string,
  but this function will not perform 
  validation."
  [input]
  (->> input
       str-to-b64-chars
       b64-chars-to-b64-ints
       b64-ints-to-raw-b64-bits
       raw-b64-bits-to-padded-b64-bits
       padded-b64-bits-to-final-b64-bits
       final-b64-bits-to-bits
       bits-to-ints
       ints-to-char-strs
       support/str-join))

;-----------------------------------------------------------------
; STEP 1 - String to [b64-characters]
;-----------------------------------------------------------------

(defn- reducer
  "Reducer function for converting
  a Base64 string to a seq of Base64
  characters-as-strings.
  
  [] \"a\" -> [\"a\"]
  [\"a\"] \"b\" -> [\"a\" \"b\"]
  [\"a\"] \"=\" -> [\"a=\"]"
  [lst cur]
  (let [tmp-init (butlast lst)
        init     (if (nil? tmp-init) [] (vec tmp-init))
        end      (last lst)
        cur-str  (str cur)]
    (if (= cur-str "=")
      (conj init (support/str-join [end cur-str]))
      (conj lst cur-str))))

(defn- str-to-b64-chars
  "Converts a Base64 string to a seq
  of Base64 characters-as-strings
  
  \"abc=\" -> [\"a\" \"b\" \"c=\"]"
  [input]
  (reduce reducer [] input))

;-----------------------------------------------------------------
; STEP 2 - [b64-characters] to [b64-integers]
;-----------------------------------------------------------------

(defn- b64-char-to-b64-int
  "Converts a Base64-character to
  a Base64-integer, adding any equals
  signs..
  
  \"Q==\" -> \"16==\""
  [input]
  (let [char-str  (support/strip-equals input)
        character (nth char-str 0)
        integer   (get b64-integer-lookup character)
        equals    (support/strip-alphanum input)]
    (str integer equals)))

(defn- b64-chars-to-b64-ints
  "Converts a seq of Base64-characters 
  to Base64-integers, adding any equals
  signs.
  
  [\"Q==\"] -> [\"16==\"]"
  [input]
  (map b64-char-to-b64-int input))

;-----------------------------------------------------------------
; STEP 3 - [b64-integers] to [unpadded b64-bit-string]
;-----------------------------------------------------------------

(defn- b64-int-to-raw-b64-bits
  "Converts a Base64-integer to a raw
  Base64-binary-string, adding any equals
  signs.
  
  \"60=\" -> \"111100=\"
  \"6\" -> \"110\""
  [input]
  (let [int-str (support/strip-equals input)
        equals  (support/strip-num input)
        integer (read-string int-str)
        bits    (Integer/toString integer 2)]
    (str bits equals)))

(defn- b64-ints-to-raw-b64-bits
  "Converts a seq of Base64-integers
  to raw Base64-binary-strings, adding
  any equals signs.
  
  [\"60=\"] -> [\"111100=\"]
  [\"6\"] -> [\"110\"]"
  [input]
  (map b64-int-to-raw-b64-bits input))

;-----------------------------------------------------------------
; STEP 4 - [unpadded b64-bit-string] to [padded b64-bit-string]
;-----------------------------------------------------------------

(defn- raw-b64-bit-seq-to-padded-b64-bit-seq
  "Pads a raw Base64-binary-string to 6
  characters, adding any equals signs.
  
  \"110\" -> \"000110\""
  [input]
  (let [bits        (support/strip-equals input)
        equals      (support/strip-num input)
        padded-bits (support/left-pad 6 bits)]
    (str padded-bits equals)))

(defn- raw-b64-bits-to-padded-b64-bits
  "Converts a seq of unpadded Base64-
  binary-strings to 6-width Base64-
  binary-strings, adding any equals signs.
  
  [\"110\"] -> [\"000110\"]"
  [input]
  (map raw-b64-bit-seq-to-padded-b64-bit-seq input))

;-----------------------------------------------------------------
; STEP 5 - [padded b64-bit-strng] to [final b64-bit-string]
;-----------------------------------------------------------------

(defn- padded-b64-bit-seq-to-final-b64-bit-seq
  "Adjusts a Base64-binary-string for any
  equals signs.
  
  \"111100=\" -> \"1111\""
  [input]
  (let [bits       (support/strip-equals input)
        equals     (support/strip-num input)
        to-drop    (* (count equals) 2)
        bit-len    (count bits)
        target-len (- bit-len to-drop)
        fin-bits   (take target-len bits)]
    (support/str-join fin-bits)))

(defn- padded-b64-bits-to-final-b64-bits
  "Adjust a seq of Base64-binary-strings
  for equals signs.
  
  [\"111100=\"] -> [\"1111\"]
  [\"111100\"] -> [\"111100\"]"
  [input]
  (map padded-b64-bit-seq-to-final-b64-bit-seq input))

;-----------------------------------------------------------------
; STEP 6 - [final b64-bit-string] to [bit string]
;-----------------------------------------------------------------

(defn- final-b64-bits-to-bits
  "Converts a seq of unpadded Base64-binary-
  strings to a seq of padded binary-strings.
  
  [\"011000\" \"01\"] -> [\"01100001\"]"
  [input]
  (support/regroup 8 input))

;-----------------------------------------------------------------
; STEP 7 - [bit string] to [integer]
;-----------------------------------------------------------------

(defn- bits-to-ints
  "Converts a seq of binary-strings
  to integers.
  
  [\"01101111\"] -> [111]"
  [input]
  (map support/bin-str-to-int input))

;-----------------------------------------------------------------
; STEP 8 - [integer] to [character-string]
;-----------------------------------------------------------------

(defn- ints-to-char-strs
  "Converts a seq of integers
  to characters-as-strings.
  
  [111] -> [\"o\"]"
  [input]
  (map (comp str char) input))